plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageService.java

```diff
@@ -28,41 +28,91 @@
 import java.net.URI;
 import java.net.URISyntaxException;
 import java.net.URL;
 import java.util.Map;
 
+import static java.util.Collections.emptyMap;
+
 public class GoogleCloudStorageService extends AbstractComponent {
 
     /** Clients settings identified by client name. */
-    private final Map<String, GoogleCloudStorageClientSettings> clientsSettings;
+    private volatile Map<String, GoogleCloudStorageClientSettings> clientsSettings = emptyMap();
+    /** Cache of client instances. Client instances are built once for each setting change. */
+    private volatile Map<String, Storage> clientsCache = emptyMap();
+
+    public GoogleCloudStorageService(final Settings settings) {
+        super(settings);
+    }
 
-    public GoogleCloudStorageService(final Environment environment, final Map<String, GoogleCloudStorageClientSettings> clientsSettings) {
-        super(environment.settings());
-        this.clientsSettings = clientsSettings;
+    /**
+     * Updates the client settings and clears the client cache. Subsequent calls to
+     * {@code GoogleCloudStorageService#client} will return new clients constructed
+     * using these passed settings.
+     *
+     * @param clientsSettings the new settings used for building clients for subsequent requests
+     * @return previous settings which have been substituted
+     */
+    public synchronized Map<String, GoogleCloudStorageClientSettings>
+            updateClientsSettings(Map<String, GoogleCloudStorageClientSettings> clientsSettings) {
+        final Map<String, GoogleCloudStorageClientSettings> prevSettings = this.clientsSettings;
+        this.clientsSettings = MapBuilder.newMapBuilder(clientsSettings).immutableMap();
+        this.clientsCache = emptyMap();
+        // clients are built lazily by {@link client(String)}
+        return prevSettings;
+    }
+
+    /**
+     * Attempts to retrieve a client from the cache. If the client does not exist it
+     * will be created from the latest settings and will populate the cache. The
+     * returned instance should not be cached by the calling code. Instead, for each
+     * use, the (possibly updated) instance should be requested by calling this
+     * method.
+     *
+     * @param clientName name of the client settings used to create the client
+     * @return a cached client storage instance that can be used to manage objects
+     *         (blobs)
+     */
+    public Storage client(final String clientName) throws Exception {
+        Storage storage = clientsCache.get(clientName);
+        if (storage != null) {
+            return storage;
+        }
+        synchronized (this) {
+            storage = clientsCache.get(clientName);
+            if (storage != null) {
+                return storage;
+            }
+            storage = SocketAccess.doPrivilegedIOException(() -> createClient(clientName));
+            clientsCache = MapBuilder.newMapBuilder(clientsCache).put(clientName, storage).immutableMap();
```

Comment:
Why not using HashMap() directly?

Reply:
Or maybe use a ConcurrentHashMap to simplify the code? You could use computeIfAbsent and the like?

Reply:
> Why not using HashMap() directly?

I suppose `MapBuilder` was created for the chaining syntax, plus making sure the result is immutable. It grew on me, don't ruin it for me now... 😛 

@jaymode 
I think ConcurrentHashMap is a behemoth we don't need.
In the common case the map contains a single entry and there is really no contention here, the "other" thread **only** clears the map. The volatile immutable map is the evolution to the volatile client reference.
If there is an occasion where a volatile reference to an immutable map trumps the concurrent map I think this must be it.
At some point (when I wished the `reinit` be blocking until all clients have been changed), I also pushed the concurrent map https://github.com/elastic/elasticsearch/pull/28517#discussion_r167814823).

Do you wish to sudo?

Reply:
> My quick glance makes this seem like it is somewhat different as there was a read write lock external to that CHM?

Yes, unfortunately the comment tied the read write lock to the CHM; I remember they were not tied though... my bad for bringing this up.

I have been thinking and I reckon this synchronized block is necessary
```
        synchronized (this) {
            storage = clientsCache.get(clientName);
            if (storage != null) {
                return storage;
            }
            storage = SocketAccess.doPrivilegedIOException(() -> createClient(clientName));
            clientsCache = MapBuilder.newMapBuilder(clientsCache).put(clientName, storage).immutableMap();
            return storage;
        }
```
because we need to keep the settings and the client references updates paired. For example, we might end up with one thread reading the settings (`clientsCache.get(clientName)`), another changing them, and then the first client will use stale settings to build a client. Now the settings don't reflect the client.
Not saying CHM will never fit, but it will not simplify this part here.